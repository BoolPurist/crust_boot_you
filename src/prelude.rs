use std::io;

pub use crate::constants;
pub use anyhow::Context;

pub type IoResult<T = ()> = Result<T, io::Error>;
pub type AppError = anyhow::Error;
pub type AppResult<T = ()> = Result<T, AppError>;

pub type ReturnToUser = AppResult<String>;
pub type PathResult = AppResult<PathBuf>;
pub type AppIoResult<T = ()> = Result<T, AppIoError>;

pub use crate::app_traits::file_manipulator::FileManipulator;
pub use crate::app_traits::path_provider::PathProvider;
pub use crate::file_management::AppIoError;
pub use crate::NotEmptyText;
pub use std::path::{Path, PathBuf};

#[cfg(test)]
pub use mockall::automock;
