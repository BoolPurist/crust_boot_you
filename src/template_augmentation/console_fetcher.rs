use super::{KeyExtact, OptAugmentationResult};

mod test_console_fetcher;
pub use test_console_fetcher::TestConsoleFetcher;

mod console_fetcher;
pub use console_fetcher::IoConsoleFetcher;

pub trait ConsoleFetcher {
    fn fetch_from(&self, key: KeyExtact<'_>) -> OptAugmentationResult;
}
