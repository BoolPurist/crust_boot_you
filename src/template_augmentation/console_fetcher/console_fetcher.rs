use std::io;

use super::ConsoleFetcher;
use crate::template_augmentation::augmentation_error::AugmentationError;
use std::io::Write;

#[derive(Default, Debug)]
pub struct IoConsoleFetcher;

impl ConsoleFetcher for IoConsoleFetcher {
    fn fetch_from(
        &self,
        key: crate::template_augmentation::KeyExtact<'_>,
    ) -> crate::template_augmentation::OptAugmentationResult {
        print!("{}: ", key);
        io::stdout().flush().map_err(|_| {
            AugmentationError::StdInProblem("Could not flush stdin after printing key as label")
        })?;
        let mut output = String::new();
        std::io::stdin()
            .read_line(&mut output)
            .map_err(|_| AugmentationError::StdInProblem("Could not read line from stdin"))?;
        let without_traling_noise = output.trim_end();
        let to_return = if without_traling_noise.trim_start().is_empty() {
            None
        } else {
            Some(without_traling_noise.to_string())
        };
        println!("");
        Ok(to_return)
    }
}
