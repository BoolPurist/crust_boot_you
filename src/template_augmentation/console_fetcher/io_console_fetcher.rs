use crate::prelude::*;
use std::io;

use super::ConsoleFetcher;
use std::io::Write;

#[derive(Default, Debug)]
pub struct IoConsoleFetcher;

impl ConsoleFetcher for IoConsoleFetcher {
    fn fetch_from(
        &self,
        key: crate::template_augmentation::KeyExtact<'_>,
    ) -> crate::template_augmentation::OptAugmentationResult {
        print!("{}: ", key);
        io::stdout()
            .flush()
            .context("Could not flush stdin after printing key as label")?;
        let mut output = String::new();
        std::io::stdin()
            .read_line(&mut output)
            .context("Could not read line from stdin")?;
        let output = output.trim_end();

        let to_return = if output.trim_start().is_empty() {
            None
        } else {
            Some(output.to_string())
        };
        println!();
        Ok(to_return)
    }
}
