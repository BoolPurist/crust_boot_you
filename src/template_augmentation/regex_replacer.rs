use crate::prelude::*;
use std::borrow::Cow;

use once_cell::sync::Lazy;
use regex::Regex;

use super::console_fetcher::TestConsoleFetcher;
use super::TestAugmentStore;

use super::{
    augmentation_error::AugmentationError,
    console_fetcher::{ConsoleFetcher, IoConsoleFetcher},
    template_extractation::TemplateExtractation,
    AugementRepository, TemplateAugmentor,
};

fn build_regex_template() -> Regex {
    Regex::new(r#"\{\{(?<value>[^\{]*)\}\}"#)
        .expect("Incorrect regex matcher for a key to be replaced within a template")
}

static REGEX_TEMPLATE: Lazy<Regex> = Lazy::new(build_regex_template);

pub struct RegexTemplateAugmentor<CF> {
    cache: AugementRepository<CF>,
}

impl Default for RegexTemplateAugmentor<IoConsoleFetcher> {
    fn default() -> Self {
        let console_fetcher = IoConsoleFetcher;
        let cache = AugementRepository::new(console_fetcher);
        Self::new(cache)
    }
}

impl<CF: ConsoleFetcher> RegexTemplateAugmentor<CF> {
    pub fn new(cache: AugementRepository<CF>) -> Self {
        Self { cache }
    }
}

impl RegexTemplateAugmentor<TestConsoleFetcher> {
    pub fn from_fake(map: TestAugmentStore) -> Self {
        let fake_console_fetcher = TestConsoleFetcher::new(map);
        let store = AugementRepository::new(fake_console_fetcher);
        Self::new(store)
    }
}

impl<CF> TemplateAugmentor for RegexTemplateAugmentor<CF>
where
    CF: ConsoleFetcher,
{
    fn try_replace<'a>(&mut self, input: &'a str) -> Result<Cow<'a, str>, AugmentationError> {
        let mut captures = REGEX_TEMPLATE.captures_iter(input).peekable();
        let needs_expansion = captures.peek().is_some();
        if needs_expansion {
            let mut expanded = String::with_capacity(input.len());
            let mut last_match = 0;
            for found in captures {
                let matched_range = found.get(0).unwrap();
                expanded.push_str(&input[last_match..matched_range.start()]);

                let extraction = {
                    let value = &found["value"];
                    let mut splited = value.split(constants::SEPERATOR_BETWEEN_DEFAULT_AND_VALUE);
                    let (key, default_value) = (splited.next().unwrap(), splited.next());
                    TemplateExtractation::FromConsole { key, default_value }
                };
                let replacement = self
                    .cache
                    .augment(&extraction)
                    .map_err(|error| AugmentationError::new(input, matched_range.start(), error))?;
                expanded.push_str(&replacement);
                last_match = matched_range.end();
            }

            expanded.push_str(&input[last_match..]);
            Ok(Cow::Owned(expanded))
        } else {
            Ok(Cow::Borrowed(input))
        }
    }
}

#[cfg(test)]
mod testing {
    use crate::template_augmentation::console_fetcher::TestConsoleFetcher;

    use super::*;

    #[test]
    fn regex_extract_one_value() {
        let input = "aa aaa {{value to get}} aaa";
        let actual = REGEX_TEMPLATE.captures(input).unwrap();
        let actual_extracted = &actual["value"];
        assert_eq!("value to get", actual_extracted);
    }
    #[test]
    fn regex_extract_two_value() {
        let input = "aa aaa {{value to get}} {{xxx}}  ";
        let map = hash_map! {
            "value to get".to_string() => "YYY".to_string(),
            "xxx".to_string() => "XXX".to_string(),
        };
        let test_augmenter = TestConsoleFetcher::new(map);

        let respo = AugementRepository::new(test_augmenter);
        let mut regex_augmentor = RegexTemplateAugmentor::new(respo);
        let actual = regex_augmentor.try_replace(&input).unwrap();

        assert_eq!("aa aaa YYY XXX  ", actual);
    }
    #[test]
    fn regex_extract_do_not_change_after_error() {
        let input = "aa aaa {{value to get}} {{!!}} {{value to get}} ";
        let map = hash_map! {
            "value to get".to_string() => "YYY".to_string(),
            "xxx".to_string() => "XXX".to_string(),
        };
        let test_augmenter = TestConsoleFetcher::new(map);

        let respo = AugementRepository::new(test_augmenter);
        let mut regex_augmentor = RegexTemplateAugmentor::new(respo);
        let _ = regex_augmentor.try_replace(&input).unwrap_err();
    }
    #[test]
    fn regex_extract_one_and_other_default_values() {
        let input = "aa aaa {{value to get}} {{!!?A}} {{!!?B}} ";
        let map = hash_map! {
            "value to get".to_string() => "YYY".to_string(),
            "xxx".to_string() => "XXX".to_string(),
        };
        let test_augmenter = TestConsoleFetcher::new(map);

        let respo = AugementRepository::new(test_augmenter);
        let mut regex_augmentor = RegexTemplateAugmentor::new(respo);
        let actual = regex_augmentor.try_replace(&input).unwrap();

        assert_eq!("aa aaa YYY A B ", actual);
    }

    #[test]
    fn regex_extract_and_no_default() {
        let input = "aa aaa {{value to get}} {{!!?A}} {{!!?B}} ";
        let map = hash_map! {
            "value to get".to_string() => "YYY".to_string(),
            "!!".to_string() => "XXX".to_string(),
        };
        let test_augmenter = TestConsoleFetcher::new(map);

        let respo = AugementRepository::new(test_augmenter);
        let mut regex_augmentor = RegexTemplateAugmentor::new(respo);
        let actual = regex_augmentor.try_replace(&input).unwrap();

        assert_eq!("aa aaa YYY XXX XXX ", actual);
    }
}
