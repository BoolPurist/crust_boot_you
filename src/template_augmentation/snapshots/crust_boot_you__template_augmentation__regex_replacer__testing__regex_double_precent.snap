---
source: src/template_augmentation/regex_replacer.rs
expression: actual
---
aa aaa {{value to get}} {{!!?A}} {{!!?B}} BBB 
                   P x!!x C  {{world}} P x!!x C
                      b
        
